src/preamble.js

```diff
@@ -134,16 +134,194 @@ var HEAP,
 var HEAP_DATA_VIEW;
 #endif
 
+function fixPointer (ptr) {
+  if (typeof ptr === 'number' && ptr < 0) {
+    return ptr >>> 0;
+  }
+  return ptr;
+}
+
+#if MEMORY64
+/**
+ * @param {string} type - Heap type
+ * @param {number} [offset] - Heap offset
+ * @param {number} [length] - typed array length
+ */
+const getHeapBlock =  (type, offset, length) => {
+  const maxSize = Math.min(wasmMemory.buffer.byteLength, 4 * 1024 * 1024 * 1024 - 8)
+  if (!offset) {
+    offset = 0
+  }
+
+  const heap = wasmMemory.buffer
+  switch (type) {
+    case 'i1':
+    case 'i8':
+      return new Int8Array(heap, offset, length || maxSize / Int8Array.BYTES_PER_ELEMENT);
+    case 'u1':
+    case 'u8':
+      return new Uint8Array(heap, offset, length || maxSize / Uint8Array.BYTES_PER_ELEMENT);
+    case 'i16':
+      return new Int16Array(heap, offset, length || maxSize / Int16Array.BYTES_PER_ELEMENT);
+    case 'u16':
+      return new Uint16Array(heap, offset, length || maxSize / Uint16Array.BYTES_PER_ELEMENT);
+    case 'i32':
+      return new Int32Array(heap, offset, length || maxSize / Int32Array.BYTES_PER_ELEMENT);
+    case 'u32':
+      return new Uint32Array(heap, offset, length || maxSize / Uint32Array.BYTES_PER_ELEMENT);
+    case 'f32':
+      return new Float32Array(heap, offset, length || maxSize / Float32Array.BYTES_PER_ELEMENT);
+    case 'f64':
+      return new Float64Array(heap, offset, length || maxSize / Float64Array.BYTES_PER_ELEMENT);
+    case 'i64':
+      return new BigInt64Array(heap, offset, length || maxSize / BigInt64Array.BYTES_PER_ELEMENT);
+    case '*':
+    case 'u64':
+      return new BigUint64Array(heap, offset, length || maxSize / BigUint64Array.BYTES_PER_ELEMENT);
+    default:
+      throw new Error('Invalid type');
+  }
+}
+#endif
+
 function updateMemoryViews() {
   var b = wasmMemory.buffer;
 #if SUPPORT_BIG_ENDIAN
   Module['HEAP_DATA_VIEW'] = HEAP_DATA_VIEW = new DataView(b);
 #endif
+#if MEMORY64
+  var maxArraySize = Math.min(b.byteLength, 4 * 1024 * 1024 * 1024 - 8);
+  var proxyHandler = (type, heapBlock) => ({
+    heapBlock,
+    copyWithin (target, start, end) {
+      target = fixPointer(target)
+      start = fixPointer(start)
+      end = fixPointer(end)
```

Comment:
I'm hoping you can simply remove `fixPointer` since all incoming pointers should already i53 (and unsigned).  See #19740.